feat(chai): updated types to v5 (ESM)

tpluscode opened the pull request:
Ok, so this is A LOT of changes. 

* [x] Update chai to v5
* [x] Preserve chai v4
* [x] Updated all usages of chai in other packages/tests
* [x] Update `sinon-chai` to v4
* [x] Updated `chai-html` to v3
* [x] Fixed `chai-json-pattern` to be accurately represented as CJS
* [x] Bumped `chai-datetime` to v1 because v0 never existed on NPM
* [ ] ~~Change all CJS packages which depend on `chai <5` or `chai ^4`~~
* [ ] ~~Kept `chai-spies` and `chai-json-schema` on `chai=^4` (see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70397#issuecomment-2336656276)~~

~~Since I updated chai to v5 which is now ESM, all CJS packages which have a dependency now need to depend on the earlier branch. I also preserved v4 of chai.~~

~~Now, I do not like adding `"chai": "<5"` to all downstream packages since most of those packages do not really depend on chai v4 and should run with v5 just fine. Unfortunately, it seems that moving chai to `devDependencies` is not an option either, unless I did something wrong.~~

~~Another issue is how types extending chai<5 use a global namespace and in the case of v5 this is now done with module augmentation.~~

~~Anyway, I did what I could to have all tests pass in all downstream packages but I'm not convinced they will all work together happily in a project.~~

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <[<url here>](https://github.com/chaijs/chai/issues/1554#issuecomment-2190191760)>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

tpluscode commented:
Ah, this requires #70174

tpluscode commented:
I've got one last problem with `chai/v4`

```
Error in chai/v4
Error: !@arethetypeswrong/cli
chai v4.3.10
@types/chai v4.3.9999
 (ignoring rules: 'no-resolution', 'cjs-only-exports-default', 'unexpected-module-syntax', 'cjs-resolves-to-esm')
🎭 Import resolved to a CommonJS type declaration file, but an ESM JavaScript file. https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/FalseCJS.md
┌───────────────────┬────────────────────────┬────────────┐
│                   │ "chai"                 │ "chai/*"   │
├───────────────────┼────────────────────────┼────────────┤
│ node10            │ 🟢                     │ (wildcard) │
├───────────────────┼────────────────────────┼────────────┤
│ node16 (from CJS) │ 🟢 (CJS)               │ (wildcard) │
├───────────────────┼────────────────────────┼────────────┤
│ node16 (from ESM) │ 🎭 Masquerading as CJS │ (wildcard) │
├───────────────────┼────────────────────────┼────────────┤
│ bundler           │ 🟢                     │ (wildcard) │
└───────────────────┴────────────────────────┴────────────┘
```

Is this because chai 4.3.10 is a dual package? What do I do about this?

jakebailey commented:
chai is known broken in https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/attw.json#L182, you'd need to add `chai/v4` now that you're introducing it.

tpluscode commented:
Thanks for pointing that out. Oh, I looked for chai in attw but after I had already removed it 🤦

andrewbranch commented:
A couple questions:

> Another issue is how types extending chai<5 use a global namespace and in the case of v5 this is now done with module augmentation.

Could v5 preserve the global namespace to minimize breaks / make upgrades easier?

> Now, I do not like adding "chai": "<5" to all downstream packages since most of those packages do not really depend on chai v4 and should run with v5 just fine. Unfortunately, it seems that moving chai to devDependencies is not an option either, unless I did something wrong.

What were the factors that prevented these packages from working with v5? I wonder if there are other options we could explore. For example, if v5 _can_ preserve the global namespace, you wouldn’t have to worry about require(esm) issues. And for test files, maybe an async import would work?

tpluscode commented:
> What were the factors that prevented these packages from working with v5?

I think the only other problem is that test files in CJS projects cannot import ESM of v5. Maybe I could also use triple slash references instead of imports in these cases.

tpluscode commented:
@andrewbranch following your advice, I kept the global `Chai` namespace in v5 and switched all but two packages to depend on any version of `@types/chai`. 

The two exceptions are `chai-json-schema` and `chai-spies` which extend the `Chai.ChaiStatic` in a way which is likely incompatible with v5 anyway ([former](https://github.com/chaijs/chai-json-schema/blob/master/index.js#L38), [latter](https://github.com/chaijs/chai-spies/blob/main/lib/spy.js#L184)). When you look at the [current implementation of `use`](https://github.com/chaijs/chai/blob/main/lib/chai.js#L31-L46), you'll find that the first argument is a new object and not the combined exports of the module. Thus, the global scope is not affected when these plugins assign additional properties.

These will still work with v5 by using the object returned from `use`

```
import { use } from 'chai'
import spies from 'chai-spies'

const chai = use(spies)

chai.spy(/* ... */)
```

cc @43081j

DangerBotOSS commented:
<!--
  0 failure: 
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-Danger;
-->



## Formatting

The following files are not formatted:
1. types/chai-xml/chai-xml-tests.ts
1. types/chai-fuzzy/chai-fuzzy-tests.ts
1. types/chai-uuid/chai-uuid-tests.ts
1. types/chai-jest-snapshot/chai-jest-snapshot-tests.ts
1. types/chai-sorted/chai-sorted-tests.ts

<details>
<summary>as well as these 30 other files...</summary>
<p>types/chai-quantifiers/chai-quantifiers-tests.ts, types/bardjs/bardjs-tests.ts, types/chai-dom/chai-dom-tests.ts, types/chai-subset/chai-subset-tests.ts, types/chai-match-pattern/chai-match-pattern-tests.ts, types/chai-json-pattern/chai-json-pattern-tests.ts, types/chai-oequal/chai-oequal-tests.ts, types/chai-like/chai-like-tests.ts, types/chai-arrays/chai-arrays-tests.ts, types/chai-jwt/chai-jwt-tests.ts, types/chai-moment/chai-moment-tests.ts, types/chai-things/chai-things-tests.ts, types/chai-datetime/chai-datetime-tests.ts, types/chai-subset-in-order/chai-subset-in-order-tests.ts, types/chai-as-promised/index.d.ts, types/chai-enzyme/chai-enzyme-tests.tsx, types/chai/index.d.ts, types/chai-string/chai-string-tests.ts, types/chai-style/chai-style-tests.ts, types/chai-roughly/chai-roughly-tests.ts, types/chai-fs/chai-fs-tests.ts, types/chai-almost/chai-almost-tests.ts, types/chai/chai-tests.ts, types/karma-chai/index.d.ts, types/sinon-chai/index.d.ts, types/deep-equal-in-any-order/deep-equal-in-any-order-tests.ts, types/dirty-chai/dirty-chai-tests.ts, types/jsx-chai/jsx-chai-tests.ts, types/ember-mocha/ember-mocha-tests.ts, types/sinon-chai/v2/sinon-chai-tests.ts</p>
</details>


Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against f2ec4bb6ddc7770490898e565e1cba82c27dfb28
</p>

KentarouTakeda commented:
I am familiar with `hexo-fs` and `hexo-util`. Changes to these are LGTM.

AviVahl reviewed (approved):
I got tagged as a maintainer... I think I've contributed one of the plugins. Anyhow, went over the changes. Impressive work. Not sure many people could have tackled this one like you did.

gitKrystan reviewed (approved):
Approved for `types/ember-mocha/ember-mocha-tests.ts`

jakebailey commented:
@tpluscode Did you verify that these packages still work with the chai v4 types? Or are all of these packages implicitly depending on v5 now?

43081j commented:
> 
> These will still work with v5 by using the object returned from `use`
> 
> ```
> import { use } from 'chai'
> import spies from 'chai-spies'
> 
> const chai = use(spies)
> 
> chai.spy(/* ... */)
> ```
> 
> cc @43081j

Just fyi this is correct. We no longer mutate the exported namespace

Let me know if you need any help with the rest of this too. I'll try read through it properly tomorrow

tpluscode commented:
> @tpluscode Did you verify that these packages still work with the chai v4 types? Or are all of these packages implicitly depending on v5 now? This is a very critical set of packages, so, safety is key.

I don't think it's necessary to check every single one, right? I shall pick some and double check.

That said, I expect that all should work with both v4 and v5 since most changes are only to switch how chai is imported in CJS tests

jakebailey commented:
chai v5 going ESM-only is very much holding people back to v4, so trying to make that keep working would be nice. Unfortunately, there isn't yet a great way to test that on DT (though with the newish DT package layout it is possible with tooling changes).

tpluscode commented:
I think that @andrewbranch's suggestion was the necessary detail needed to ensure backwards compatibility.

And many popular packages have already become ESM-only so I think chai is already in the clear and I don't expect much resistance in switching. Especially that the plugin ecosystem is not directly depending on the module syntax of chai itself.

tpluscode commented:
@jakebailey I created a test project using some the WIP v5 types: https://github.com/tminuscode/chai-5-types

I found a small mistake in the `use` function and fixed it, so that's good.

More interestingly, I decided that I should include the types for `chai-spies` and `chai-json-schema` in the migration. After all, when using these packages with chai@v5, the usage will have to be modified anyway, as discussed above. But nothing in the new types prevents these packages to function as expected. I updated their respective tests to reflect that too

tpluscode commented:
There is, however, one small detail, again relating to `chai-spies` and `chai-json-schema`. Their (recommended?) usage in v4 is like

```ts
import chai from 'chai'
import chaiSpies from 'chai-spies'

chai.use(chaiSpies)

// find `.spy` on the chai export
const spy = chai.spy(() => {});
```

This will stop working because `chai.use` does not mutate but, more importantly, because v5 does not have a default export. From the types perspective, it would be simple to add something like this to the types

```ts
declare const legacy: Chai.ChaiStatic
export default legacy
```

But at this point we'd be introducing something that does not exist in the real package. @43081j, would you be find with adding a default export like this to v5 to reduce breaking changes when migrating from 4 to 5?

43081j commented:
People currently have to change from a default import to upgrade to v5 right? So that is already a known upgrade cost

Are you concerned there's too many dependents to update to use named imports?

tpluscode commented:
> People currently have to change from a default import to upgrade to v5 right? So that is already a known upgrade cost

Fair enough. Than we should be good, and when updating to chai v5 they would discover that usage of `chai-spies` and `chai-json-schema` needs to change.

> Are you concerned there's too many dependents to update to use named imports?

I get a feeling that there is an edge case which we can tackle. If existing code uses chai v4 and cannot update but install types of v5, that would result in issues but that probably is not something worth addressing. Maybe I'm chasing my own tail now...

43081j commented:
Yeah I think if someone is still using v4 and updates to v5 types, that is a mistake. Not sure the best way we can prevent people from doing that though (just hope they understand semver?)

Most plugins should work with v5 but consumers usage would have to change (i.e you should still be able to pass the plugin in fine without the plugin authors changing anything, but you pass it in/use it differently)

jakebailey commented on types/chai-datetime/package.json:
```diff
@@ -1,7 +1,7 @@
     "private": true,
     "name": "@types/chai-datetime",
-    "version": "0.0.9999",
+    "version": "1.0.9999",
```
This library is actually at 1.8; not sure if it's valuable to fix that now or not. 1.0 is okay.

tpluscode replied:
I think I did not go all the way to 1.8 because I did not have the time to see if anything unrelated might have changed

jakebailey commented on types/chai-datetime/chai-datetime-tests.ts:
```diff
@@ -1,9 +1,9 @@
-const assert = chai.assert;
+import("chai").then(chai => chai.use(chaiDateTime));
+
+const expect: Chai.ExpectStatic = {} as any;
+const assert: Chai.AssertStatic = {} as any;
```
For all of these `const expect: ...`, I assume these are here because they are in CJS packages which cannot import chai anymore?

I would actually suggest writing these like:

```ts
declare const expect: Chai.ExpectStatic;
declare const assert: Chai.AssertStatic;
```

Rather than using the cast. I personally find the intent clearer for DT tests where declaring things sort of means "I don't care I'm just declaring this to write the tests".

jakebailey reviewed (commented):
Gave this a full look through; I think my only commentary is a nit on how tests declare `expect`/`assert`.

Other than that, all of the tests are consistent. But, they all seem to do  `import("chai").then(chai => chai.use(...));`.

I assume that this is actually _impossible_ to do within CJS consumers of chai, right? Or that chai just cannot be used in CJS contexts anymore? Such that these lines are simply to satisfy the tests in some way given the packages themselves are all pretty much just CJS?

tpluscode commented:
> Gave this a full look through; I think my only commentary is a nit on how tests declare `expect`/`assert`.

Changed those as proposed in the inline comment.
 
> Other than that, all of the tests are consistent. But, they all seem to do `import("chai").then(chai => chai.use(...));`.
> 
> I assume that this is actually _impossible_ to do within CJS consumers of chai, right?

Quite the opposite. CJS project should be able to use chai v5 in this exact way, importing it dynamically to get register plugins, or get a hold of `expect` and `assert`.

> Or that chai just cannot be used in CJS contexts anymore? Such that these lines are simply to satisfy the tests in some way given the packages themselves are all pretty much just CJS?

So, while I don't think top-level await is allowed in tests so I could wrap each test in the async callback like

```diff
-declare const expect: Chai.ExpectStatic;
-declare const assert: Chai.AssertStatic;

import('chai').then(({ use, expect, assert } => {
  use(testedPlugin)

  // continue with tests using expect and assert
})
```

I figured that this will add even more noise to an already busy PR without really adding any value to the plugins' tests. They only need to test that the plugin satisfies the signature to `use` and extend the the global interface of `expect`, `assert`, etc.

jakebailey commented:
Okay, so it's just that you have to move all of your tests into an async function. That's fine; I think I was just confused because it definitely wouldn't work to just fire off a promise at the top and then start declaring things, I think, which is what these tests did.

jakebailey reviewed (approved):
Alright, let's see how this goes.

The only other concern is just that ATA will start pulling this in people's editors if they don't have `@types/chai` installed themselves and are writing JS, but my understanding is that the actual shape of the API is the same (so editors using bundler resolution are unlikely to even notice a difference in the unlikely case they use chai v4 but load v5 types).

tpluscode commented:
Ready to merge
